framework/src/org/checkerframework/framework/util/AnnotatedTypes.java

```diff
@@ -884,26 +885,34 @@ public static boolean hasTypeQualifierElementTypes(ElementType[] elements, Class
      *       as bounds of type variables and wildcards.
      * </ol>
      */
-    public static boolean isValidType(
+    public static Result isValidType(
```

Comment:
How about keeping the return type here `boolean` and doing:

```java
Result isValid = scanner.visit(type);
return isValid.isSuccess();
```

That would hide this implementation detail and wouldn't require the checker-framework-inference change.

Reply:
Ah, sorry. I ignored that the caller uses the error code in `checker.report(isValid, tree);`.
Would it be ugly to report the error here?

Reply:
The `BaseTypeValidator` also directly reports errors, so I think this would be OK. Maybe add a note to the javadoc of this method.

Reply:
If the error is reported here, then checker-framework-inference would report the error.

Reply:
Good point.
Could we just move all of `AnnotatedTypes.isValidType` into the `TypeValidator`/`BaseTypeValidator`?
Logically, they belong together so putting one validity check into a static method seems wrong.
Then the `BaseTypeVisitor` code above would simply become:

```java
     public boolean validateType(Tree tree, AnnotatedTypeMirror type) {
         return typeValidator.isValid(type, tree);
     }
```

And inference already has it's own `TypeValidator`.